README.md
# ZXSpectrum-Basicode
Source code of BASICODE-3 for the ZX Spectrum computer

This project documents my implementation of BASICODE-3 for the ZX Spectrum computer, written in 1986-1987.
For an explanation of BASICODE, please read https://github.com/janbredenbeek/basicode.

The source code is divided into six parts:

- BASICODE standard subroutines and support subroutines (BASIC)
- Read/Write/translate routines
- 42 column screen routines
- Extra BASIC functions
- BEXT Extended Basic interpreter part 1
- BEXT Extended Basic interpreter part 2

Unless otherwise stated, all parts are written in Z80 Assembly. The original code was written for the Hisoft Devpac assembler and, because of tight memory constraints, hardly documented. No use has been made of macros and very little use of special assembler directives so it should not be too difficult to port this to another platform.

I don't own a ZX Spectrum anymore but have successfully tested the program on various emulators such as Spectaculator and EightyOne, despite the huge dependency on the original ZX Spectrum ROM and accurate CPU timing. As for the latter, Spectaculator is my preferred choice. It can also read in BASICODE programs saved as 8-bit .WAV files at high speed. You must set it to emulate the 48K Spectrum though as the program does not currently work in 128K mode.


History.md
# BASICODE on the ZX Spectrum - A Story

Let me introduce myself. My name is Jan Bredenbeek from the Netherlands. In 1983, I bought a Sinclair ZX Spectrum. At that time, I was also aware of the Hobbyscoop radio programme with its BASICODE-2 transmissions. So I asked the NOS if there was a BASICODE translator program available for the Spectrum. It turned out that there wasn't one (of course I should have bought a C64 instead - just joking :)). However there was a guy named Rik Koevoets working on a ZX81 translator, to be aired on Christmas Day that year. And despite the fact that a ZX81 is not a Spectrum, I did have a ZX81-to-Spectrum conversion program, typed in from pages of listings in the Your Computer magazine, so I thought that would finally enable me to load the BASICODE programs I had recorded from the radio.

Well, eventually Christmas Day 1983 came and my Christmas present arrived. Well, sort of. I spent nearly all of the Christmas holidays trying to run the ZX81 translator on my Spectrum. Eventually I succeeded, but it still needed a lot of tweaks to just read in BASICODE programs, let alone to run them. Nevertheless I regard those days as sweet memories. I did learn a lot about Z80 Assembly language programming, and also a lot about ZX Spectrum machine code disassembling and hacking ;-)).

A couple of months later, the 'official' NOS BASICODE-2 book and cassette arrived, including a ZX Spectrum translator also from Rik Koevoets. But to me it appeared somewhat as a disappointment. Writing a BASICODE translator for the Sinclair machines is not an easy task given the incompatibilities of ZX Basic, but on the Spectrum almost no BASICODE program would run without manual intervention. There was the 'one-letter variable name' issue so a name like IN$ was 'illegal', like line numbers above 9999. The Spectrum had only a 32-column display where BASICODE expected 40. And when you did a DIM A$(10), you would get a 'string array' of 10 single characters - and always 10, since ZX Basic treated the last subscript as a 'length' parameter! Similarly, LEFT$ MID$ and RIGHT$ were missing (but there was a quick workaround by defining them as FN L$/M$/R$). All in all, It was so tough that even the BBC decided to bring out a modified BASICODE to be able to run it on the Sinclair machines - much to the discontent of the original Dutch inventors.

As 1984 progressed, more Spectrum enthousiasts embarked on writing a better Basicode translator. One of them was Herman Ament, who wrote a more user-friendly program that included code to allow for a 42-column display and automatic replacement of 'illegal' line numbers and variable names. But still, the incompatibilities between Sinclair and BASICODE (read: Microsoft) Basic proved to be a barrier I wanted to get rid of. So... I decided to hack Sinclair's Basic to make it compatible with BASICODE!

Nowadays, this would seem to be an impossible task given the complexity of software and copyright issues. But in the Spectrum era there was a book called 'The Complete Spectrum ROM Disassembly' by Ian Logan, which was considered as THE bible to ZX Spectrum machine code programming. Also, all Spectrums sold were equipped with the same ROM version so programmers could happily use the ROM routines without having to bother about versions (at least at that time). Needless to say there were already a lot of 'extended Basics' available for the Spectrum, but none of them addressed the issues faced with BASICODE. So I had to do it myself...

By 1985, I was already in contact with Hans G. Janssen from the NOS Hobbyscoop team. He was aware of the challenges Spectrum users faced with Basicode and welcomed my attempt to make a 'translator program to end all previous translator programs' for the Spectrum with enthousiasm. Eventually, it took me about 5 months to write it (as a student I had a bit of spare time then...) and in June 1985 it finally aired. It was also included in the new edition of the BASICODE-2 book and cassette (for obscure reasons named BASICODE-2A), which arrived just in time for the HCC dagen (then Holland's biggest computer fair) in November 1985.

But work wasn't finished, and the BASICODE-3 standard was already under way. In the year that followed I did a complete rewrite of the Spectrum translator, adding the graphics and file management features and making the Basic faster. Sadly enough, the launch of BASICODE-3 in June 1986 was troubled by the conflict between the NOS and the BASICODE Foundation over copyright issues, resulting in a split-up. The NOS continued to air programs in the old BASICODE-2 standard while the Foundation found a new broadcast partner in the TROS, which started to air BASICODE-3 programs in July 1986.


BC3-v31.BAS
    1 REM : RANDOMIZE USR VAL "59453"
   10 GO SUB 60: GO TO 1000
   20 IF A<100 THEN LET A=100
   30 CLEAR ,A
   40 BORDER 0: PAPER 0: INK 7: GO SUB 100: LET HO=41: LET VE=23: LET SV=7: LET HG=256: LET VG=192
   50 GO SUB 60: RANDOMIZE : GO TO 1010
   60 DEF FN LEFT$(A$,N)=USR 58840: DEF FN MID$(A$,N,N)=USR 58843: DEF FN RIGHT$(A$,N)=USR 58846: DEF FN SHIFT$(A$)=USR 58855: DEF FN EOF(N)=USR 58858: DEF FN o()=USR VAL "59474": DEF FN o$()=USR VAL "59477": RETURN 
  100 CLS : PRINT AT 0,0;: RETURN 
  110 PRINT AT INT VE AND VE>=0 AND VE<24-PEEK 56806,INT HO AND HO>=0 AND HO<42;: RETURN 
  120 LET HO=43-PEEK 56802: LET VE=24-PEEK 56803: IF HO>41 THEN LET HO=0: LET VE=VE+1
  125 IF VE>23-PEEK 56806 THEN LET VE=23-PEEK 56806
  130 RETURN 
  150 PRINT "   ";SR$;"   ";: RETURN 
  200 LET IN=USR 56793
  201 LET IN$=CHR$ IN AND IN
  202 LET IN=IN-(32 AND IN>95 AND IN<127): RETURN 
  210 LET IN$=INKEY$#2: IF IN$="" THEN GO TO 210
  211 LET IN=CODE IN$: GO TO 202
  220 LET IN=CODE SCREEN$ (VE,HO): GO TO 202
  250 BEEP .25,21: RETURN 
  260 LET RV=RND: RETURN 
  270 LET FR=USR 59456: RETURN 
  280 POKE 65367,FR: RETURN 
  300 LET SR$=STR$ SR: RETURN 
  310 LET OS=ABS SR+.5*10^-CN: LET OI=INT OS: LET OD=OS-OI+1
  311 LET SR$=""
  312 IF OS>=1E8 THEN GO TO 321
  313 IF NOT CN THEN LET OD$="": GO TO 317
  315 LET OD$="."+FN MID$(STR$ OD,3,CN)
  316 IF LEN OD$<CN+1 THEN LET OD$=OD$+"0": GO TO 316
  317 LET SR$=STR$ OI+OD$
  318 IF SR<0 AND VAL SR$ THEN LET SR$="-"+SR$
  319 IF LEN SR$<CT THEN LET SR$=" "+SR$: GO TO 319
  320 IF LEN SR$>CT THEN LET SR$=""
  321 IF LEN SR$<CT THEN LET SR$="*"+SR$: GO TO 321
  322 RETURN 
  330 LET SR$=FN SHIFT$(SR$): RETURN 
  350 LPRINT SR$;: RETURN 
  360 LPRINT : RETURN 
  400 IF SV AND SD<105 THEN BEEP SD/10,SP-60: RETURN 
  401 LET OS=SD: LET OI=IN: LET OI$=IN$
  402 GO SUB 450: IF SD THEN GO TO 402
  403 LET SD=OS: LET IN=OI: LET IN$=OI$: RETURN 
  450 LET SD=(SD*5+USR 56796)/5: LET IN=PEEK 23560: GO TO 201
  500 POKE 65281+(NF+4)*4,0: LET IN=0: IF NF<2 THEN LET O=550 AND USR 59459: LET O=NF+4 AND USR (53950+3*NF): GO TO 543
  510 OPEN #FN o()+NF+4;"m";FN o()+1;FN o$()+NF$
  530 RETURN 
  540 LET O=550 AND USR 59459
  541 IF FN EOF(NF+4) THEN LET IN=1: LET IN$="": GO TO 543
  542 INPUT #NF+4;IN$: LET IN=FN EOF(NF+4)
  543 LET O=0 AND USR 59459: RETURN 
  550 PRINT #0;AT 1,0;"                                ";: LET IN=-1: LET IN$=""
  555 RETURN 
  560 PRINT #NF+4;SR$: LET IN=0: RETURN 
  580 LET IN=0: IF NF<2 THEN LET O=NF+4 AND USR 53956: RETURN 
  590 CLOSE #FN o()+NF+4
  595 RETURN 
  600 GO SUB 100: POKE 23678,191: RETURN 
  620 PLOT INVERSE NOT NOT CN;INT (256*(HO-INT HO)),191-INT (192*(VE-INT VE)): RETURN 
  630 DRAW INVERSE NOT NOT CN;INT (256*(HO-INT HO))-PEEK 23677,191-INT (192*(VE-INT VE))-PEEK 23678: RETURN 
  650 LET OH=INT (256*(HO-INT HO)): LET OV=191-INT (192*(VE-INT VE))
  660 POKE 23677,OH: POKE 23678,OV: INVERSE NOT NOT CN: LET O$=SR$ AND USR 56799: POKE 23677,OH: POKE 23678,OV: INVERSE 0: RETURN 
  950 POKE 65367,0: LET NF=0
  951 IF PEEK 23635+256*PEEK 23636-PEEK 23631-256*PEEK 23632<33 THEN GO TO 953
  952 GO SUB 580: LET NF=NF+1: IF NF<8 THEN GO TO 951
  953 CLEAR : STOP 
  960 CLS : LET a$="basicode-3"
  961 PRINT "CASSETTE OR MICRODRIVE/DISK ?";
  962 GO SUB 210
  963 IF IN=77 THEN GO TO 970
  964 IF IN<>67 THEN GO TO 962
  965 SAVE A$ LINE 966: POKE PEEK 23631+256*PEEK 23632+2,181: SAVE "CODE BASICODE 3"CODE 53950,11418: STOP 
  966 CLEAR VAL "53949": LOAD ""CODE : RANDOMIZE USR VAL "59450": RANDOMIZE USR 56790: GO TO 1000
  970 SAVE *"m";FN o()+1;FN o$()+A$ LINE FN o()+971: SAVE *"m";FN o()+1;"CODE BASICODE 3"CODE FN o()+53950,FN o()+11418: STOP 
  971 CLEAR VAL "53949": LOAD *"m";SGN PI;"CODE BASICODE 3"CODE : RANDOMIZE USR VAL "59450": RANDOMIZE USR 56790: GO TO 1000
  980 CLS : INPUT "PROGRAM NAME: ";A$: IF NOT LEN A$ OR LEN A$>10 THEN GO TO 980
  981 GO TO 961
 1000 BORDER 0: PAPER 0: INK 7: CLS : PRINT "         ** B A S I C O D E  3 **"
 1010 PRINT '''"  VERTAALPROGRAMMA / TRANSLATION PROGRAM"
 1020 PRINT ''"               VERSION 3.1"
 1030 PRINT '''"      VOOR MENU TOETS '*' & <ENTER>"
 1040 PRINT '''"            FOR MENU ENTER '*'"
 1050 PRINT '''"  (C) 1986, 87 JAN BREDENBEEK, HILVERSUM"


BCFN.ASM
; BASICODE-3 v3.1 FOR THE ZX SPECTRUM 48K
; COPYRIGHT (C) 1985-1987 BY JAN BREDENBEEK
; RELEASED UNDER GNU v3 PUBLIC LICENSE, 2017

; FUNCTIONS FOR BASICode EXTension
; REVISION 1.0: 10 DECEMBER 1985
; REVISION 3.1: 24 NOVEMBER 1987

       ORG    58840
	   
; Entry points for functions from BASIC
	   
       JP     LEFT$           ; LEFT$
       JP     MID$            ; MID$
       JP     RIGHT$          ; RIGHT$
       JP     VAL             ; VAL
       JP     SCREEN$         ; SCREEN$
       JP     SHIFT$          ; SHIFT$
       JP     EOF             ; EOF
       JP     VALSUB          ; VAL subroutine
       JP     VALFRC          ; VAL fractions

CHARS  EQU    58072		; pointer to character set (6-bit)

; Stack arguments for FN LEFT$, MID$, RIGHT$

STK_ARG
       LD     HL,(23563)      ; DEFADD holds addr of arguments area
       INC    HL              ; (like VARS area, but local to FN)
       INC    HL              ; point to string argument
       CALL   #33B4           ; stack string parameter
       INC    HL
       INC    HL              ; point to first numeric argument
STK_A1 CALL   #33B4           ; stack this one too
       CALL   #1E99           ; FIND-INT2
       PUSH   BC              ; save value
       CALL   #2BF1           ; STK-FETCH
       
; At this point DE points to the address of the string argument and BC holds
; its length (see explanation in BEXT.TXT). The numeric argument is now tested
; against the string length and adjusted if necessary (it cannot be greater)
       
       LD     H,B             ; String length to HL
       LD     L,C
       POP    BC              ; Get N
       AND    A
       SBC    HL,BC
       ADD    HL,BC           ; test if N>=LEN(A$)
       EX     DE,HL           ; length to DE, pointer to HL
       RET    NC              ; return if N<=LEN(A$)
       LD     B,D             ; else, adjust N
       LD     C,E
       RET

; LEFT$, MID$ and RIGHT$ enter here. They are called as user-defined (FN) 
; functions from BASIC, pointing to USR addresses (we'll tweak the type of
; the returned value to string before returning).
; Note that FN MID$ can have two or three arguments: MID$(A$,N) returns
; the equivalent of A$(N TO) and MID$(A$,N,M) returns A$(N TO N+M-1).
; The BEXT interpreter was changed in v3 to allow for a 'missing parameter' by
; putting all FFs in the 5 variable locations of the 'missing' argument, so
; we can pick this up here.
       
LEFT$  CALL   STK_ARG         ; LEFT$; get parameters and jump to store result
       JR     STK_STO

MID$   CALL   STK_ARG         ; MID$; get A$ and N first
       JR     NC,NOTEMPTY     ; if N<LEN(A$)
       LD     BC,#0000        ; else, string will be empty so adjust BC
       JR     STK_STO         ; ... and stack it
NOTEMPTY
       DEC    HL
       ADD    HL,BC           ; point to first char of substring
       EX     DE,HL           ; HL now holds LEN(A$), DE=pointer
       INC    HL
       LD     A,B             ; N may not be zero
       OR     C
       JP     Z,#2A20         ; else, error "Subscript wrong"
       SBC    HL,BC           ; now get length of substring
       LD     B,H             ; BC holds new length
       LD     C,L
       CALL   #2AB6           ; now stack the result string
       
; We still need to consider the form MID$(A$,N,M) 
       
       LD     HL,(23563)      ; DEFADD
       LD     BC,16
       ADD    HL,BC           ; point to location where M is stored
       LD     A,(HL)
       INC    A               ; if M is not given, A holds $FF
       JR     NZ,MID$_2       ; jump if M is present
       LD     DE,#FFFF        ; else, signal 'whole substring'
       CALL   #2D8C           ; and store dummy length
MID$_2 CALL   STK_A1          ; now stack M (or dummy '65535')
       JR     STK_STO         ; jump to store substring

; RIGHT$(A$,N) returns the rightmost N chars of A$ (ie A$(LEN(A$)-N+1 TO))
       
RIGHT$ CALL   STK_ARG         ; get A$ and N
       ADD    HL,DE           ; point to 1 past string
       AND    A
       SBC    HL,BC           ; and go back N chars (N<=LEN(A$))
       
; All three functions eventually come here to store the result string.
; HL points to start of (sub)string, BC holds length.
       
STK_STO
       EX     DE,HL           ; pointer to DE
       CALL   #2AB2           ; STK-STO-$ stacks string result
       
; We now need to do some mangling since the string functions were called by USR
; and the expression evaluator expects a numeric result. First we drop
; the return address of STACK-BC (since we already have stacked our string) and
; then we modify the type of the stacked result (pushed on the machine stack 
; from the expression evaluator) from 'numeric' to 'string'.
       
       POP    BC              ; drop address of STACK-BC
       LD     HL,4
       ADD    HL,SP           ; skipover return addresses from calculator
       RES    7,(HL)          ; this signals 'string result' when DE gets
                              ; popped from the stack at $2764
       JP     #35BF           ; exit via STACK-PNTRS 

; FN SHIFT$(A$) converts a string to uppercase       
       
SHIFT$ LD     HL,(23563)      ; DEFADD
       XOR    A
       CPIR                   ; look for end of parameter name (null byte)
       CALL   #33B4           ; stack string parameter
       CALL   #2BF1           ; STK-FETCH
       PUSH   BC              ; save length
       PUSH   DE              ; and pointer
       INC    BC              ; add extra byte (in case length is 0)
       RST    #30             ; and allocate this in WORKSP
       POP    HL              ; get pointer to source
       PUSH   DE              ; save pointer to destination
       LDIR                   ; and copy over the string
       POP    DE              ; restore destination
       POP    BC              ; original length
       PUSH   BC
       PUSH   DE
       
; now copy over the string, converting to uppercase as we go       
       
SH_LP  LD     A,B
       OR     C
       JR     Z,SH_END
       LD     A,(DE)
       CP     #60
       JR     C,SH_NXT
       CP     #80
       JR     NC,SH_NXT
       AND    #DF             ; note: braces, tilde etc get also converted!
       LD     (DE),A
SH_NXT INC    DE
       DEC    BC
       JR     SH_LP           ; loop while BC<>0
SH_END POP    HL              ; get destination pointer
       POP    BC              ; and length
       JR     STK_STO         ; and store result

; This is called by the expression evaluator (via jump vector) for the VAL
; function. This evaluates the string argument, which must represent a valid
; number. Thus it's not as sophisticated as the Spectrum's original VAL 
; function, which allowed complete expressions to be evaluated. However, in
; BASICODE, this is not part of the spec and even undesirable as it may lead
; to stopping the program with "Nonsense in BASIC" when trying to evaluate 
; something like VAL ("2abc"). In BASICODE, it is defined to return zero when
; the argument does not represent a valid number (sorry fans - after all it
; was designed around M$ Basic...)
       
VAL    CALL   #2BF1           ; STK-FETCH
       INC    BC
       PUSH   DE
       RST    #30             ; allocate len+1 in WORKSP
       POP    HL
       PUSH   DE
       LDIR                   ; and copy over
       EX     DE,HL
       DEC    HL
       LD     (HL),#0D        ; and padd with CR byte
       LD     HL,(23645)      ; get CH-ADD
       EX     (SP),HL         ; and swap with string address
       LD     (23645),HL
       CALL   VALSUB          ; now do actual operation
       POP    HL
       LD     (23645),HL      ; restore original CH-ADD
       RET
       
; This subroutine does the actual VAL operation
       
VALSUB RST    #28             ; invoke calculator
       DEFB   #A1;STK_ONE
       DEFB   #C2;ST_MEM_2
       DEFB   #02;DELETE
       DEFB   #38;END_CALC    ; MEM-2 now preset to 1
       RST    #18             ; get current char
       CP     "+"
       JR     Z,S_DONE        ; skipover '+'
       CP     "-"             
       JR     NZ,V_DEC
       DEC    (IY+100)        ; if '-', set MEM-2 to 0
S_DONE RST    #20             ; get next char
V_DEC  CALL   #2D3B           ; read integer part
VALFRC CP     "."             ; decimal point?
       JR     NZ,E_FORMAT     ; no, look for E-format
DEC_STO                       ; else, preload MEM-0 with 1
       RST    #28
       DEFB   #A1;STK_ONE
       DEFB   #C0;ST_MEM_0
       DEFB   #02;DELETE
       DEFB   #38;END_CALC
NXTDGT CALL   #0074           ; loop for fractional digits
       CALL   #2D22           ; stack new digit (if any)
       JR     C,E_FORMAT      ; until non-digit found
       RST    #28
       DEFB   #E0;GET_MEM_0
       DEFB   #A4;STK_TEN
       DEFB   #05;DIVISION
       DEFB   #C0;ST_MEM_0    ; 1 in MEM-0 goes to 0.1, 0.01 etc for each turn
       DEFB   #04;MULTIPLY    ; and multiply this by value of new digit
       DEFB   #0F;ADDITION    ; and add to running value
       DEFB   #38;END_CALC
       JR     NXTDGT
E_FORMAT                      ; consider xx.yyEzz format
       CP     "e"
       JR     Z,S_FLAG
       CP     "E"
       JR     NZ,VAL_END      ; if neither e or E we are finished
S_FLAG LD     B,#FF           ; preset B to $FF
       CALL   #0074
       CP     "+"
       JR     Z,SIGN_DONE
       CP     "-"             ; is exponent negative?
       JR     NZ,ST_E
       INC    B               ; yes, set B to 0
SIGN_DONE
       CALL   #0074           ; and skipover '+' or '-' after E
ST_E   PUSH   BC
       CALL   #2D3B           ; INT-TO-FP
       CALL   #2DD5           ; FP-TO-A
       POP    BC              ; restore BC
       JP     C,#31AD         ; overflow if exponent > 255
       AND    A
       JP     M,#31AD         ; and if >127 (in practice, anything above about
                              ; 38 will overflow later)
       INC    B               ; test sign, -1 for positive, 0 for negative
       JR     Z,E_CALC
       NEG                    ; if sign negative, negate A
E_CALC CALL   #2D4F           ; call E-TO-FP to calculate value in E-format
VAL_END                       ; almost finished, but finally look at sign
       RST    #28
       DEFB   #E2;GET_MEM_2   ; ... which is stored in MEM-2
       DEFB   #00;JUMP_TRUE   ; if positive, it will be 1
       DEFB   #02
       DEFB   #1B;NEGATE      ; else, negate result
       DEFB   #38;END_CALC    ; finished
       RET
       
; This is called by the SCREEN$ function. Since our BASICODE characters are
; only 6 bits wide to make 42 columns on the screen possible, we also have to
; write our own modified SCREEN$ to recognise the narrower characters...
; (it's called by GOSUB 220 from the standard subroutines)
       
SCREEN$
       CALL   #2307           ; STK-TO-BC
       LD     A,D
       OR     E
       AND    #80             ; one or both arguments negative?
       JR     NZ,SCR_EMPTY    ; yes, return empty string
       LD     A,B             ; B holds column number
       CP     42              ; .. which must be 0 to 41
       JR     NC,SCR_EMPTY
       LD     A,C             ; C holds row number
       CP     24              ; .. which must be 0 to 23
       JR     NC,SCR_EMPTY
       RRCA                   ; divide row by 8
       RRCA
       RRCA
       AND    #E0
       LD     L,A             ; L = 32 * (row MOD 8) (1)
       LD     A,C
       AND    #18
       OR     #40
       LD     H,A             ; H = 64 + 8 * (row DIV 8)
       LD     A,B
       ADD    A,A
       ADD    A,B
       LD     B,A             ; B = column * 3
       RRA
       SRL    A               ; A = (column * 3) DIV 4 (2)
       OR     L
       LD     L,A             ; Now L = (1) + (2), the low address byte
       LD     A,B
       ADD    A,A
       AND    #07
       INC    A               ; A = (column * 6) MOD 8 + 1
       LD     (23655),A       ; store in BREG as shift counter for SC_MATCH
       LD     DE,CHARS        ; pointer to our own charset (not the Spectrum's)
       LD     B,#60           ; loop for at most 96 characters
       
; Now enter a loop to find a matching character
; B = counter (from 96 to 1)
; DE = pointer to current entry in character set bitmap
; HL = current address in display file
       
SCR_LP PUSH   BC
       PUSH   HL
       PUSH   DE
       CALL   SC_MATCH        ; Match bitmap against character line
       JR     Z,MATCH         ; if match, loop for another 7 lines
       CP     #FC
       JR     NZ,SC_NXT       ; also consider case of INVERSE 1
MATCH  LD     C,A             ; save result (either 0 or $FC)
       LD     B,7             ; look for another 7 lines
SC_ROW INC    H               ; next display line is 256 bytes away
       INC    E               ; but next line in character set only 1
       CALL   SC_MATCH        ; now match them up
       XOR    C               ; and include mask from first line
       JR     NZ,SC_NXT       ; no match, next one
       DJNZ   SC_ROWS         ; loop for 7 lines
       POP    BC              ; we've found one!
       POP    BC
       POP    BC              ; B = original counter
       LD     A,#80
       SUB    B               ; form ASCII code in A
       LD     BC,#0001
       RST    #30
       LD     (DE),A          ; store 1 character in WORKSP
       RET                    ; finished
SC_NXT POP    DE              ; no match, get original char pointer
       LD     HL,#0008
       ADD    HL,DE
       EX     DE,HL           ; bump DE by 8
       POP    HL              ; original screen address
       POP    BC              ; original counter in B
       DJNZ   SCR_LP          ; loop for 96 characters
SCR_EMPTY
       LD     BC,#0000        ; no luck, return empty string
       RET
       
; This subroutine tries the actual match. The 6 bits found at the display
; position are picked up, rotated if needed and matched against the bitmap
; from the character set.
; Note that in our 42 column display it is possible for one character to 
; occupy 2 display locations (4 columns occupy exactly 3 locations), unlike
; the Spectrum's normal 32 column display where 1 character is exactly 1 byte.
; Hence, some bit munging is necessary to get the bits into the right place.
; The number of shifts needed + 1 is already stored in the BREG variable.
       
SC_MATCH
       PUSH   HL              ; save screen address
       LD     A,(HL)
       INC    HL
       LD     L,(HL)
       LD     H,A             ; get 2 screen bytes in HL
       LD     A,(23655)       ; get number of shifts + 1
SC_M_L DEC    A               ; count number of shifts
       JR     Z,SC_XOR        ; done
       ADD    HL,HL           ; else, shift bits left
       JR     SC_M_LP         ; and loop back
SC_XOR LD     A,(DE)          ; now match against character bitmap
       XOR    H               ; (only 6 msb's of HL)
       POP    HL              ; restore screen address
       AND    #FC             ; mask off unwanted bits
       RET

; FN EOF(channel) returns EOF status of an opened channel
; This is done by reading a byte from the channel and trapping the EOF error
; unfortunately, any character successfully read now has to be 'pushed back'
; into the stream and this is done by temporarily saving this character in a
; table and changing the address of the input routine to code which will return
; this character on the next call (and then restore the original pointer).
       
EOF    LD     HL,(23563)      ; get DEFADD
       XOR    A
       CPIR                   ; find start of argument value
       CALL   #33B4           ; store on calculator stack
       CALL   #1E94           ; fetch this into A
       CP     #10             
       JP     NC,#1725        ; 'invalid stream' if 16 or higher
       
; A table located at address $FF01 holds EOF status information for each
; stream. Each entry has 4 bytes:
; byte 0: status; 0 = unused; $FF = input pending
; byte 1: character to return on next input call
; byte 2 and 3: address of original input routine       
       
       LD     L,A
       ADD    HL,HL
       ADD    HL,HL           ; multiply stream # by 4
       INC    L
       LD     H,#FF           ; form table entry address in HL
       LD     BC,0            ; preset BC to 0 for 'no EOF'
       BIT    7,(HL)          ; test 'used' flag
       RET    NZ              ; return 'no EOF' when input is pending
       PUSH   HL              ; save table address
       LD     HL,(23633)
       PUSH   HL              ; save CURCHL pointer
       CALL   #1601           ; select this stream
       EXX                    ; go to alternate register set
       
; this code changes the error stack pointer to IN_RET below and tries to fetch
; a character from the selected channel. After restoring the original ERR_SP,
; the error return is tested for EOF.
       
       PUSH   HL              ; save original HL
       LD     HL,(23613)
       PUSH   HL              ; save original ERR_SP
       LD     HL,IN_RET
       PUSH   HL
       LD     (23613),SP      ; make IN_RET the error return point
       CALL   #15DE           ; now call INPUT_AD1 to fetch a character
       POP    HL              ; No error - drop IN_RET address
IN_RET POP    HL              ; any error returns here
       LD     (23613),HL      ; restore original ERR_SP
       POP    HL              ; original HL
       EXX                    ; restore original register set
       LD     HL,(23633)      ; CURCHL
       EX     (SP),HL
       LD     B,A             ; save character read
       CALL   #1615           ; reset CURCHL to original value
       POP    DE              ; get pointer to channel where we've read from
       POP    HL              ; table entry address
       LD     A,(23610)
       INC    A               ; test error code
       JR     Z,NO_EOF        ; no error so we have to save character read
       CP     #08
       JP     NZ,#0058        ; if error was other than EOF, report it
       LD     (IY+0),#FF      ; else, reset EOF error
       LD     BC,1            ; and return 1 for EOF
       RET
       
; A character has been read from the input, but it has to be pushed back into
; the input stream. This is done by saving it into the table and temporarily
; changing the input routine's address to our own routine which returns it
; on the next input call.       
       
NO_EOF LD     (HL),#FF        ; flag 'input pending'
       INC    HL
       LD     (HL),B          ; store character to be returned
       INC    HL
       EX     DE,HL           ; HL points to channel info
       INC    HL
       INC    HL              ; step past output routine pointer
       LDI                    
       LDI                    ; copy input routine pointer to table entry
       LD     DE,NEWINP       ; our temporary input routine
       DEC    HL
       LD     (HL),D
       DEC    HL
       LD     (HL),E          ; enter this into the channel's input routine ptr
       LD     BC,0            ; no EOF
       RET
       
; A temporary input routine is used to return the character read from the call
; to FN EOF(). After this, the address of the original input routine is
; restored.
       
NEWINP LD     HL,(23633)      ; CURCHL
       LD     DE,(23631)      ; CHANS
       XOR    A
       SBC    HL,DE
       LD     B,H
       LD     C,L
       INC    BC              ; BC = offset of channel from CHANS + 1
       LD     HL,#5C16        ; STRMS
EOF_LP LD     E,(HL)          ; now we have to find the stream number attached
       INC    HL              ; to our channel
       LD     D,(HL)
       INC    HL
       EX     DE,HL
       AND    A
       SBC    HL,BC           ; offset in STRMS matches our channel?
       JR     Z,EOFFND        ; yes, look for pending input
       EX     DE,HL
       INC    A
       CP     #10
       JR     C,EOF_LP        ; loop for streams 0 to 15
       AND    A               ; if not found, return C and Z reset ('no input')
       RET
EOFFND RLCA                   
       RLCA
       INC    A
       LD     L,A
       LD     H,#FF           ; find table entry
       LD     DE,(23633)      ; CURCHL
       INC    DE
       INC    DE              ; step past output routine pointer
       LD     (HL),0          ; flag 'no pending input'
       INC    HL
       LD     A,(HL)          ; and fetch character to be returned
       INC    HL              ; point to original routine pointer
       LDI                    
       LDI                    ; copy this to channel pointer
       SCF                    ; signal 'succesful read'
       RET
